src/data/complex/is_R_or_C.lean

```diff
@@ -49,21 +50,51 @@ class is_R_or_C (K : Type*) extends nondiscrete_normed_field K, normed_algebra 
 namespace is_R_or_C
 
 variables {K : Type*} [is_R_or_C K]
-local notation `𝓚` := @is_R_or_C.of_real K _
 local postfix `†`:100 := @is_R_or_C.conj K _
 
-lemma of_real_alg : ∀ x : ℝ, 𝓚 x = x • (1 : K) :=
-λ x, by rw [←mul_one (𝓚 x), smul_coe_mul_ax]
+/--
+Coercions that go from a concrete structure such as
+`ℝ` to an arbitrary field `K` should be set up as follows:
+```lean
+@[priority 900] instance : has_coe_t ℝ K := ⟨...⟩
+```
+
+It needs to be `has_coe_t` instead of `has_coe` because otherwise type-class
+inference would loop when constructing the transitive coercion `ℝ → ℝ → ℝ → ...`.
+The reduced priority is necessary so that it doesn't conflict with instances
+such as `has_coe_t α (option α)`.
+
+For this to work, we reduce the priority of the `coe_base` and `coe_trans`
+instances because we want the instances for `has_coe_t` to be tried in the
+following order:
+
+ 1. `has_coe_t` instances declared in mathlib (such as `has_coe_t α (with_top α)`, etc.)
+ 2. `coe_base`
+ 3. `is_R_or_C.cast_coe : has_coe_t ℝ K` etc.
+ 4. `coe_trans`
+
+If `coe_trans` is tried first, then `is_R_or_C.cast_coe` doesn't get a chance to apply.
+-/
+attribute [instance, priority 950] coe_base
+attribute [instance, priority 500] coe_trans
```

Comment:
Since these are quite central to lean, they should probably be adjusted in a much more fundamental file - otherwise people might get surprising behavior changes when they import this file.

Should this change go into lean itself?

Reply:
Doesn't the namespace restrict the scope of these lines? I copied this from `data/nat/cast` where it's also namespaced.

Reply:
Only if you write `local attribute`:

```lean
class foo (α : Type*) := (x : α)
instance nat.foo_0 : foo ℕ := ⟨0⟩
instance nat.foo_37 : foo ℕ := ⟨37⟩

def foo1 : foo ℕ := infer_instance
example : foo1 = nat.foo_37 := rfl

namespace baz

local attribute [instance, priority 950] nat.foo_0
local attribute [instance, priority 500] nat.foo_37

end baz

def foo2 : foo ℕ := infer_instance
example : foo2 = nat.foo_37 := rfl

namespace baz

attribute [instance, priority 950] nat.foo_0
attribute [instance, priority 500] nat.foo_37

end baz

def foo3 : foo ℕ := infer_instance
example : foo3 = nat.foo_0 := rfl
```

Reply:
I see -- thanks!